File: MinimumGeneticMutation3.py
class Solution:
    def minMutation(self, start, end, bank):
        """
        :type start: str
        :type end: str
        :type bank: List[str]
        :rtype: int
        """
        def canTranslate(s1, s2):
            count = 0
            for i in range(len(s1)):
                if s1[i] != s2[i]:
                    count += 1
                if count > 1:
                    return False
            return count == 1

        def minMutation(start, end):
            if start == end:
                return 0
            step = len(bank) + 1
            for i, gens in enumerate(bank):
                if not visited[i] and canTranslate(gens, start):
                    visited[i] = True
                    temp = minMutation(gens, end)
                    if temp != -1:
                        step = min(step, temp)
                    visited[i] = False
            return -1 if step == len(bank) + 1 else step + 1

        visited = [False] * len(bank)
        return minMutation(start, end)


if __name__ == '__main__':
    start = input()
    end = input()
    bank = [gens for gens in input().split()]
    solution = Solution()
    print(solution.minMutation(start, end, bank))

File: MinimumGeneticMutation3.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<unordered_set>

using namespace std;

class Solution {
public:
    int minMutation(string start, string end, vector<string>& bank) {
        
        if(bank.empty()) return -1;
        queue<string> q{{start}};
        unordered_set<string> s(bank.begin(), bank.end());
        unordered_set<string> visited;
        vector<char> gens{'A', 'G', 'C', 'T'};
        
        int step = 0;
        while(!q.empty()) {
        	int size = q.size();
        	for(int i = 0; i < size; i++) {
        		string t = q.front();q.pop();
        		visited.insert(t);
        		if(t == end) return step;
        		for(int j = 0; j < t.size(); j++) {
        			char old = t[j];
        			for(char gen : gens)
        				if(gen != old) {
        					t[j] = gen;
        					if(s.find(t) != s.end() && visited.find(t) == visited.end()) q.push(t);
						}
					t[j] = old;
				}
			}
			step++;
		}
        
        return -1;
    }
};

int main() {
	
	string start, end;
	cin>>start;
	cin>>end;
	
	vector<string> bank;
	int n;
	cin>>n;
	
	for(int i = 0; i < n; i++) {
		string s;
		cin>>s;
		bank.push_back(s);
	}
	
	Solution *solution = new Solution();
	cout<<solution->minMutation(start, end, bank);
	
	return 0;
	
}
